### PR TITLE
[persist] Don't assume old parts are totally consolidated

### PR DESCRIPTION
**Pull request by bkirwi**
### Motivation

Support for #21615.

In particular, this is necessary if we want to assume that the output of the `Consolidator` is indeed consolidated.

### Tips for reviewer

To my knowledge the last important bug was fixed in df46c7464df110fe6cdd97564becb000c0d08933, and 0.67.0 is the first release that includes it.

I think this means it's safe to assert on ordering in the consolidator now, but I'll run all the upgrade tests to make sure.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/persist-client/src/iter.rs [hunk 1]
```diff
@@ -610,21 +643,31 @@ where
...
+                                Ordering::Equal => {
+                                    let (_, _, _, d1) = part
+                                        .pop(&mut self.parts)
+                                        .expect("popping from a non-empty iterator");
+                                    d0.plus_equals(&d1);
+                                }
+                                Ordering::Greater => {
+                                    // Don't want to log the entire KV, but it's interesting to know
+                                    // whether it's KVs going backwards or 'just' timestamps.
+                                    panic!("data arrived at the consolidator out of order (kvs equal? {})", (*k0, *v0) == (*k1, *v1));
```
Comment (danhhz) [inline src/persist-client/src/iter.rs:665]: do we want to put this behind a feature flag at first, at least until we're relying on it for correctness? it'd be unfortunate if the release process finds some weird edge case that CI didn't catch and we have to put out a patch deploy

File: src/persist-client/src/internal/paths.rs [hunk 2]
```diff
@@ -124,6 +124,17 @@ impl PartialBatchKey {
         PartialBatchKey(format!("{}/{}", version, part_id))
     }
 
+    pub fn split(&self) -> (WriterKey, PartId) {
+        let (writer_key, part_id) = self
+            .0
+            .split_once('/')
+            .expect("partial batch key should contain a /");
+
+        let writer_key = WriterKey::from_str(writer_key).expect("correctly formatted writer key");
```
Comment (danhhz) [inline src/persist-client/src/internal/paths.rs:133]: won't this panic if it encounters a key written by an old version of persist?

also, this overlaps a bit with `BlobKey::parse_ids`, it'd be nice to dedup that if it's not too gnarly

File: src/persist-client/src/iter.rs (hunk 1 above)
Comment (bkirwi) [reply src/persist-client/src/iter.rs:665]: Every _use_ of this code is behind a feature flag, so I feel relatively comfortable here. Does that sound like a reasonable safeguard or should we add a flag for just the assert?

File: src/persist-client/src/internal/paths.rs (hunk 2 above)
Comment (bkirwi) [reply src/persist-client/src/internal/paths.rs:133]: `WriterKey` has variants for both old-style and new-style keys, so this should parse successfully in both cases. (And the newer keys sort larger than all old-style keys, so the check below should also be correct.

Good call on the duplication - will see what I can do!

**Review by danhhz — APPROVED** [review]
lgtm!

would love to see the de-dent happen. deduping WriterKey parsing logic i'll leave to your judgement

**Comment by bkirwi** [comment]
Landed both - thanks for the review!

**Pull request merged**